File: yelp-scraper/browser/python/yelp.py
"""Yelp scraper using the official Scrapeless Python SDK + Playwright over CDP.
the function names and emitted field shapes match verbatim, so downstream
code can Three public entry points (mirror the upstream reference):
- `scrape_pages(urls)` -> list of business profile dicts
- `scrape_reviews(url, max_reviews=None)` -> list of reviews from Yelp's GraphQL feed
- `scrape_search(keyword, location, max_pages=None)` -> list of search result props
"""

from __future__ import annotations

import asyncio
import base64
import json
import math
import os
from typing import Any, Dict, List, Optional
from urllib.parse import urlencode

import jmespath
from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "US"
DEFAULT_SESSION_TTL = 180

# ---------------------------------------------------------------------------
# Scrapeless plumbing
# ---------------------------------------------------------------------------

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

def _looks_blocked(html: str) -> bool:
    if not html or len(html) < 5000:
        return True
    if "captcha-delivery.com" in html or "DataDome CAPTCHA" in html:
        return True
    if "px-captcha" in html or "Pardon Our Interruption" in html:
        return True
    return False


async def _fetch_rendered_html(
    url: str,
    ready_selector: Optional[str] = None,
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
    retries: int = 2,
    warmup: bool = True,
) -> str:
    """Mint a cloud browser session, navigate, wait for a stable marker, return HTML."""
    last_error: Optional[Exception] = None
    for attempt in range(retries + 1):
        client = _client()
        session = client.browser.create(
            ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
        )
        ws = session.browser_ws_endpoint
        async with async_playwright() as pw:
            browser = await pw.chromium.connect_over_cdp(ws)
            try:
                page = await browser.new_page()
                if warmup:
                    # Hit yelp.com first so the session picks up DataDome cookies
                    # and second-hop traffic clears anti-bot more reliably.
                    try:
                        await page.goto("https://www.yelp.com/", wait_until="domcontentloaded", timeout=30000)
                        await asyncio.sleep(1.5)
                    except Exception:
                        pass
                await page.goto(url, wait_until="domcontentloaded", timeout=45000)
                if ready_selector:
                    try:
                        await page.wait_for_selector(ready_selector, timeout=15000)
                    except Exception as e:
                        logger.warning("wait_for_selector failed (continuing): {}", e)
                html = await page.content()
                if html and not _looks_blocked(html):
                    return html
                last_error = RuntimeError("anti-bot block" if _looks_blocked(html) else "empty HTML")
            except Exception as e:  # noqa: BLE001
                last_error = e
                logger.warning("attempt {} failed: {}", attempt + 1, e)
            finally:
                try:
                    await browser.close()
                except Exception:
                    pass
    raise RuntimeError(f"giving up after {retries + 1} attempts: {last_error}")

async def _post_json(
    url: str,
    payload: str,
    headers: Dict[str, str],
    *,
    proxy_country: str = DEFAULT_PROXY_COUNTRY,
) -> str:
    """POST a JSON body via the cloud browser (so we inherit anti-bot session state)."""
    client = _client()
    session = client.browser.create(
        ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
    )
    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp(session.browser_ws_endpoint)
        try:
            ctx = await browser.new_context()
            # Warm session cookies on www.yelp.com so the GQL endpoint accepts us.
            page = await ctx.new_page()
            referer = headers.get("referer", "https://www.yelp.com/")
            try:
                await page.goto(referer, wait_until="domcontentloaded", timeout=30000)
            except Exception:
                pass
            response = await ctx.request.post(url, data=payload, headers=headers)
            return await response.text()
        finally:
            try:
                await browser.close()
            except Exception:
                pass

# ---------------------------------------------------------------------------
# Parsers — mirror the upstream reference's function names verbatim
# ---------------------------------------------------------------------------

_BUSINESS_LD_TYPES = {
    "Restaurant", "LocalBusiness", "FoodEstablishment", "Bar",
    "CafeOrCoffeeShop", "Hotel", "Store", "Organization",
}


def _read_business_jsonld(sel: Selector) -> Optional[Dict[str, Any]]:
    candidates: List[Dict[str, Any]] = []
    for raw in sel.xpath('//script[@type="application/ld+json"]/text()').getall():
        try:
            data = json.loads(raw)
        except Exception:
            continue
        if isinstance(data, list):
            candidates.extend([d for d in data if isinstance(d, dict)])
        elif isinstance(data, dict):
            candidates.append(data)
    for node in candidates:
        t = node.get("@type")
        if isinstance(t, str) and t in _BUSINESS_LD_TYPES:
            return node
        if isinstance(t, list) and any(x in _BUSINESS_LD_TYPES for x in t):
            return node
    return None


def _address_to_string(addr: Any) -> str:
    if isinstance(addr, str):
        return addr
    if not isinstance(addr, dict):
        return ""
    parts = [
        addr.get("streetAddress"),
        addr.get("addressLocality"),
        addr.get("addressRegion"),
        addr.get("postalCode"),
        addr.get("addressCountry"),
    ]
    return ", ".join(str(p).strip() for p in parts if isinstance(p, str) and p.strip())


def _jsonld_open_hours(node: Dict[str, Any]) -> Dict[str, str]:
    out: Dict[str, str] = {}
    spec = node.get("openingHoursSpecification")
    if not isinstance(spec, list):
        return out
    for entry in spec:
        if not isinstance(entry, dict):
            continue
        days = entry.get("dayOfWeek")
        if not isinstance(days, list):
            days = [days]
        opens = entry.get("opens", "")
        closes = entry.get("closes", "")
        rng = f"{opens}-{closes}" if opens and closes else (opens or closes or "")
        for day in days:
            if not isinstance(day, str):
                continue
            key = day.rsplit("/", 1)[-1][:3].lower()
            out[key] = rng
    return out


def parse_page(html: str) -> Dict[str, Any]:
    """Parse business data from a yelp business profile HTML."""
    sel = Selector(text=html)
    xp = lambda q: (sel.xpath(q).get(default="") or "").strip()
    open_hours: Dict[str, str] = {}
    for day in sel.xpath('//th/p[contains(@class,"day-of-the-week")]'):
        name = (day.xpath("text()").get() or "").strip()
        value = (day.xpath("../following-sibling::td//p/text()").get() or "").strip()
        if name:
            open_hours[name.lower()] = value
    out = dict(
        name=xp("//h1/text()"),
        website=xp('//p[contains(text(),"Business website")]/following-sibling::p/a/text()'),
        phone=xp('//p[contains(text(),"Phone number")]/following-sibling::p/text()'),
        address=xp('//a[contains(text(),"Get Directions")]/../following-sibling::p/text()'),
        logo=xp('//img[contains(@class,"businessLogo")]/@src'),
        claim_status="".join(sel.xpath('//span[span[contains(@class,"claim")]]/text()').getall()).strip().lower(),
        open_hours=open_hours,
    )

    # JSON-LD fallback — DOM selectors are fragile across Yelp's A/B layouts.
    business = _read_business_jsonld(sel)
    if business:
        if not out["name"]:
            out["name"] = str(business.get("name", "")).replace("&apos;", "'")
        if not out["phone"]:
            out["phone"] = str(business.get("telephone", ""))
        if not out["address"]:
            out["address"] = _address_to_string(business.get("address"))
        if not out["website"]:
            same = business.get("sameAs")
            if isinstance(same, list) and same:
                out["website"] = same[0]
            elif isinstance(business.get("url"), str) and "yelp.com" not in business["url"]:
                out["website"] = business["url"]
        if not out["logo"]:
            img = business.get("image")
            if isinstance(img, str):
                out["logo"] = img
            elif isinstance(img, dict) and isinstance(img.get("url"), str):
                out["logo"] = img["url"]
        if not out["open_hours"]:
            oh = _jsonld_open_hours(business)
            if oh:
                out["open_hours"] = oh
    return out

def parse_business_id(html: str) -> Optional[str]:
    """Pluck the encoded business id from the page meta tag."""
    sel = Selector(text=html)
    return sel.css('meta[name="yelp-biz-id"]::attr(content)').get()

def parse_review_data(response_text: str) -> Dict[str, Any]:
    """Parse the GraphQL batch response from `/gql/batch`."""
    data = json.loads(response_text)
    reviews = data[0]["data"]["business"]["reviews"]["edges"]
    parsed_reviews = []
    for review in reviews:
        result = jmespath.search(
            """{
            encid: encid,
            text: text.{full: full, language: language},
            rating: rating,
            feedback: feedback.{coolCount: coolCount, funnyCount: funnyCount, usefulCount: usefulCount},
            author: author.{encid: encid, displayName: displayName, displayLocation: displayLocation, reviewCount: reviewCount, friendCount: friendCount, businessPhotoCount: businessPhotoCount},
            business: business.{encid: encid, alias: alias, name: name},
            createdAt: createdAt.utcDateTime,
            businessPhotos: businessPhotos[].{encid: encid, photoUrl: photoUrl.url, caption: caption, helpfulCount: helpfulCount},
            businessVideos: businessVideos,
            availableReactions: availableReactionsContainer.availableReactions[].{displayText: displayText, reactionType: reactionType, count: count}
            }""",
            review["node"],
        )
        parsed_reviews.append(result)
    total_reviews = data[0]["data"]["business"]["reviewCount"]
    return {"reviews": parsed_reviews, "total_reviews": total_reviews}

def parse_search(html: str) -> Dict[str, Any]:
    """Parse listing data from the inline `react-root-props` JSON."""
    search_data: List[Dict[str, Any]] = []
    total_results = 0
    sel = Selector(text=html)
    script = sel.xpath("//script[@data-id='react-root-props']/text()").get()
    if not script:
        return {"search_data": search_data, "total_results": total_results}
    data = json.loads(script.split("react_root_props = ")[-1].rsplit(";", 1)[0])
    search_page_props = (
        data.get("legacyProps", {}).get("searchAppProps", {}).get("searchPageProps", {})
    )
    if search_page_props:
        total_results = search_page_props.get("paginationInfo", {}).get("totalResults", 0)
    main_content = search_page_props.get("mainContentComponentsListProps", [])
    for item in main_content:
        if item.get("bizId") and item.get("searchResultBusiness") is not None:
            search_data.append(item)
    return {"search_data": search_data, "total_results": total_results}

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference's exports verbatim
# ---------------------------------------------------------------------------

async def scrape_pages(urls: List[str]) -> List[Dict[str, Any]]:
    """Scrape Yelp business profile pages."""
    result: List[Dict[str, Any]] = []
    for url in urls:
        html = await _fetch_rendered_html(url, ready_selector="h1")
        result.append(parse_page(html))
    logger.success(f"scraped {len(result)} business pages")
    return result

async def _request_reviews_api(url: str, start_index: int, business_id: str) -> str:
    """POST Yelp's GetBusinessReviewFeed GQL batch endpoint."""
    pagination_data = json.dumps({"version": 1, "type": "offset", "offset": start_index})
    after = base64.b64encode(pagination_data.encode("utf-8")).decode("utf-8")

    payload = json.dumps(
        [
            {
                "operationName": "GetBusinessReviewFeed",
                "variables": {
                    "encBizId": business_id,
                    "reviewsPerPage": 10,
                    "selectedReviewEncId": "",
                    "hasSelectedReview": False,
                    "sortBy": "DATE_DESC",
                    "languageCode": "en",
                    "ratings": [5, 4, 3, 2, 1],
                    "isSearching": False,
                    "after": after,
                    "isTranslating": False,
                    "translateLanguageCode": "en",
                    "reactionsSourceFlow": "businessPageReviewSection",
                    "minConfidenceLevel": "HIGH_CONFIDENCE",
                    "highlightType": "",
                    "highlightIdentifier": "",
                    "isHighlighting": False,
                },
                "extensions": {
                    "operationType": "query",
                    "documentId": "ef51f33d1b0eccc958dddbf6cde15739c48b34637a00ebe316441031d4bf7681",
                },
            }
        ]
    )
    headers = {
        "accept": "*/*",
        "accept-language": "en-US,en;q=0.9",
        "cache-control": "no-cache",
        "content-type": "application/json",
        "origin": "https://www.yelp.com",
        "referer": url,
        "x-apollo-operation-name": "GetBusinessReviewFeed",
    }
    return await _post_json("https://www.yelp.com/gql/batch", payload, headers)

async def scrape_reviews(url: str, max_reviews: Optional[int] = None) -> List[Dict[str, Any]]:
    """Scrape reviews for a Yelp business via the GraphQL feed."""
    logger.info("scraping the business id from the business page")
    business_html = await _fetch_rendered_html(url, ready_selector='meta[name="yelp-biz-id"]')
    business_id = parse_business_id(business_html)
    if not business_id:
        logger.error("could not find business id on {}", url)
        return []

    logger.info("scraping the first review page")
    first = await _request_reviews_api(url=url, business_id=business_id, start_index=1)
    review_data = parse_review_data(first)
    reviews = review_data["reviews"]
    total_reviews = review_data["total_reviews"]

    if max_reviews and max_reviews < total_reviews:
        total_reviews = max_reviews

    logger.info("scraping review pagination, remaining ({}) more pages", total_reviews // 10)
    for offset in range(11, total_reviews, 10):
        try:
            resp = await _request_reviews_api(url=url, business_id=business_id, start_index=offset)
            reviews.extend(parse_review_data(resp)["reviews"])
        except Exception as e:  # noqa: BLE001
            logger.error("error scraping review page offset {}: {}", offset, e)
    logger.success(f"scraped {len(reviews)} reviews from review pages")
    return reviews

async def scrape_search(
    keyword: str, location: str, max_pages: Optional[int] = None
) -> List[Dict[str, Any]]:
    """Scrape Yelp search results for `keyword` near `location`."""

    def make_search_url(offset: int) -> str:
        base = "https://www.yelp.com/search?"
        return base + urlencode({"find_desc": keyword, "find_loc": location, "start": offset})

    logger.info("scraping the first search page")
    first_html = await _fetch_rendered_html(
        make_search_url(0), ready_selector="script[data-id='react-root-props']"
    )
    data = parse_search(first_html)
    search_data = data["search_data"]
    total_results = data["total_results"]

    total_pages = max(1, math.ceil(total_results / 10))
    if max_pages and max_pages < total_pages:
        total_pages = max_pages

    logger.info("scraping search pagination, remaining ({}) more pages", total_pages - 1)
    for offset in range(10, total_pages * 10, 10):
        try:
            html = await _fetch_rendered_html(
                make_search_url(offset), ready_selector="script[data-id='react-root-props']"
            )
            search_data.extend(parse_search(html)["search_data"])
        except Exception as e:  # noqa: BLE001
            logger.error("error scraping search page offset {}: {}", offset, e)
    logger.success(f"scraped {len(search_data)} listings from search pages")
    return search_data

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    return obj

File: yelp-scraper/browser/nodejs/yelp.mjs
// Yelp scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// the function names and emitted field shapes match verbatim, so downstream
// code can import { Scrapeless } from "@scrapeless-ai/sdk";
import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";

const DEFAULT_PROXY_COUNTRY = "US";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

function looksBlocked(html) {
  if (!html || html.length < 5000) return true;
  if (/captcha-delivery\.com|DataDome CAPTCHA|px-captcha|Pardon Our Interruption/i.test(html)) return true;
  return false;
}

async function fetchRenderedHtml(url, readySelector, { proxyCountry = DEFAULT_PROXY_COUNTRY, retries = 2, warmup = true } = {}) {
  let lastError;
  for (let attempt = 0; attempt <= retries; attempt++) {
    const { browserWSEndpoint } = await client().browser.create({
      proxyCountry,
      sessionTTL: DEFAULT_SESSION_TTL,
    });
    let browser;
    try {
      browser = await puppeteer.connect({ browserWSEndpoint });
      const page = await browser.newPage();
      // Warmup: visit yelp.com first so the session picks up DataDome cookies
      // before navigating to the deeper URL — second-hop traffic is far less
      // likely to trip the CAPTCHA interstitial.
      if (warmup) {
        try {
          await page.goto("https://www.yelp.com/", { waitUntil: "domcontentloaded", timeout: 30000 });
          await new Promise((r) => setTimeout(r, 1500));
        } catch (_) {}
      }
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
      if (readySelector) {
        try {
          await page.waitForSelector(readySelector, { timeout: 15000 });
        } catch (_) {
          // non-fatal
        }
      }
      const html = await page.content();
      if (html && !looksBlocked(html)) return html;
      lastError = new Error(looksBlocked(html) ? "anti-bot block" : "empty HTML");
    } catch (e) {
      lastError = e;
    } finally {
      try { await browser?.close(); } catch (_) {}
    }
  }
  throw new Error(`giving up after ${retries + 1} attempts: ${lastError?.message}`);
}

async function postJson(url, payload, headers, { proxyCountry = DEFAULT_PROXY_COUNTRY } = {}) {
  const { browserWSEndpoint } = await client().browser.create({
    proxyCountry,
    sessionTTL: DEFAULT_SESSION_TTL,
  });
  let browser;
  try {
    browser = await puppeteer.connect({ browserWSEndpoint });
    const page = await browser.newPage();
    const referer = headers.referer ?? "https://www.yelp.com/";
    try {
      await page.goto(referer, { waitUntil: "domcontentloaded", timeout: 30000 });
    } catch (_) {}
    const text = await page.evaluate(
      async ([u, body, h]) => {
        const r = await fetch(u, { method: "POST", headers: h, body, credentials: "include" });
        return await r.text();
      },
      [url, payload, headers],
    );
    return text;
  } finally {
    try { await browser?.close(); } catch (_) {}
  }
}

// ---------------- parsers (mirror the upstream reference's function names) ----------------

function _readBusinessJsonLd($) {
  const blocks = [];
  $('script[type="application/ld+json"]').each((_, el) => {
    try {
      const raw = $(el).contents().text();
      if (!raw) return;
      const data = JSON.parse(raw);
      if (Array.isArray(data)) blocks.push(...data);
      else blocks.push(data);
    } catch (_) {}
  });
  const types = new Set([
    "Restaurant", "LocalBusiness", "FoodEstablishment", "Bar",
    "CafeOrCoffeeShop", "Hotel", "Store", "Organization",
  ]);
  const isBusiness = (node) => {
    if (!node || typeof node !== "object") return false;
    const t = node["@type"];
    if (typeof t === "string") return types.has(t);
    if (Array.isArray(t)) return t.some((x) => types.has(x));
    return false;
  };
  return blocks.find(isBusiness) || null;
}

function _addressToString(addr) {
  if (!addr || typeof addr !== "object") return typeof addr === "string" ? addr : "";
  const parts = [addr.streetAddress, addr.addressLocality, addr.addressRegion, addr.postalCode, addr.addressCountry]
    .filter((x) => typeof x === "string" && x.trim());
  return parts.join(", ");
}

function _jsonLdOpenHours(node) {
  // schema.org openingHoursSpecification → { mon: "08:00-17:00", ... }
  const out = {};
  const spec = node?.openingHoursSpecification;
  if (!Array.isArray(spec)) return out;
  for (const entry of spec) {
    const days = Array.isArray(entry?.dayOfWeek) ? entry.dayOfWeek : [entry?.dayOfWeek];
    const opens = entry?.opens ?? "";
    const closes = entry?.closes ?? "";
    const range = opens && closes ? `${opens}-${closes}` : (opens || closes || "");
    for (const day of days) {
      if (typeof day !== "string") continue;
      const key = day.split("/").pop().slice(0, 3).toLowerCase();
      out[key] = range;
    }
  }
  return out;
}

export function parsePage(html) {
  const $ = cheerio.load(html);
  const openHours = {};
  $('th p').each((_, el) => {
    const $p = $(el);
    if (!/day-of-the-week/.test($p.attr("class") || "")) return;
    const name = $p.text().trim();
    const value = $p.parent().parent().find("td p").first().text().trim();
    if (name) openHours[name.toLowerCase()] = value;
  });

  const xText = (sel) => ($(sel).first().text() || "").trim();
  const out = {
    name: xText("h1"),
    website: xText('p:contains("Business website") + p a'),
    phone: xText('p:contains("Phone number") + p'),
    address: (() => {
      const link = $('a:contains("Get Directions")').first();
      return (link.parent().next("p").text() || "").trim();
    })(),
    logo: $('img[class*="businessLogo"]').first().attr("src") || "",
    claim_status: $('span:has(> span[class*="claim"])').first().text().trim().toLowerCase(),
    open_hours: openHours,
  };

  // JSON-LD fallback — the DOM selectors above are fragile across Yelp's
  // A/B layouts. The Restaurant/LocalBusiness JSON-LD block reliably carries
  // name, telephone, address, logo and openingHoursSpecification.
  const business = _readBusinessJsonLd($);
  if (business) {
    if (!out.name) out.name = (business.name ?? "").toString().replace(/&apos;/g, "'");
    if (!out.phone) out.phone = (business.telephone ?? "").toString();
    if (!out.address) out.address = _addressToString(business.address);
    if (!out.website) {
      // Yelp sometimes carries a sameAs[] or url
      const same = business.sameAs;
      if (Array.isArray(same) && same.length) out.website = same[0];
      else if (typeof business.url === "string" && !business.url.includes("yelp.com")) out.website = business.url;
    }
    if (!out.logo) {
      const img = business.image;
      if (typeof img === "string") out.logo = img;
      else if (img && typeof img === "object" && typeof img.url === "string") out.logo = img.url;
    }
    if (!Object.keys(out.open_hours).length) {
      const oh = _jsonLdOpenHours(business);
      if (Object.keys(oh).length) out.open_hours = oh;
    }
  }
  return out;
}

export function parseBusinessId(html) {
  const $ = cheerio.load(html);
  return $('meta[name="yelp-biz-id"]').attr("content") || null;
}

function pluck(node, path) {
  // tiny safe getter, e.g. pluck(node, "feedback.coolCount")
  return path.split(".").reduce((acc, k) => (acc == null ? acc : acc[k]), node);
}

export function parseReviewData(responseText) {
  const data = JSON.parse(responseText);
  const reviews = data[0].data.business.reviews.edges;
  const parsedReviews = reviews.map((edge) => {
    const node = edge.node;
    return {
      encid: node.encid,
      text: { full: pluck(node, "text.full"), language: pluck(node, "text.language") },
      rating: node.rating,
      feedback: {
        coolCount: pluck(node, "feedback.coolCount"),
        funnyCount: pluck(node, "feedback.funnyCount"),
        usefulCount: pluck(node, "feedback.usefulCount"),
      },
      author: {
        encid: pluck(node, "author.encid"),
        displayName: pluck(node, "author.displayName"),
        displayLocation: pluck(node, "author.displayLocation"),
        reviewCount: pluck(node, "author.reviewCount"),
        friendCount: pluck(node, "author.friendCount"),
        businessPhotoCount: pluck(node, "author.businessPhotoCount"),
      },
      business: {
        encid: pluck(node, "business.encid"),
        alias: pluck(node, "business.alias"),
        name: pluck(node, "business.name"),
      },
      createdAt: pluck(node, "createdAt.utcDateTime"),
      businessPhotos: (node.businessPhotos ?? []).map((p) => ({
        encid: p.encid,
        photoUrl: pluck(p, "photoUrl.url"),
        caption: p.caption,
        helpfulCount: p.helpfulCount,
      })),
      businessVideos: node.businessVideos,
      availableReactions: (pluck(node, "availableReactionsContainer.availableReactions") ?? []).map((r) => ({
        displayText: r.displayText,
        reactionType: r.reactionType,
        count: r.count,
      })),
    };
  });
  const totalReviews = data[0].data.business.reviewCount;
  return { reviews: parsedReviews, total_reviews: totalReviews };
}

export function parseSearch(html) {
  const $ = cheerio.load(html);
  const script = $("script[data-id='react-root-props']").first().contents().text();
  const searchData = [];
  let totalResults = 0;
  if (!script) return { search_data: searchData, total_results: totalResults };
  const raw = script.split("react_root_props = ").slice(-1)[0];
  const trimmed = raw.replace(/;\s*$/, "");
  const data = JSON.parse(trimmed);
  const props = data?.legacyProps?.searchAppProps?.searchPageProps ?? {};
  totalResults = props?.paginationInfo?.totalResults ?? 0;
  for (const item of props?.mainContentComponentsListProps ?? []) {
    if (item?.bizId && item?.searchResultBusiness != null) searchData.push(item);
  }
  return { search_data: searchData, total_results: totalResults };
}

// ---------------- scrape functions ----------------

export async function scrapePages(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchRenderedHtml(url, "h1");
    out.push(parsePage(html));
  }
  return out;
}

async function requestReviewsApi(url, startIndex, businessId) {
  const paginationData = JSON.stringify({ version: 1, type: "offset", offset: startIndex });
  const after = Buffer.from(paginationData, "utf-8").toString("base64");
  const payload = JSON.stringify([
    {
      operationName: "GetBusinessReviewFeed",
      variables: {
        encBizId: businessId,
        reviewsPerPage: 10,
        selectedReviewEncId: "",
        hasSelectedReview: false,
        sortBy: "DATE_DESC",
        languageCode: "en",
        ratings: [5, 4, 3, 2, 1],
        isSearching: false,
        after,
        isTranslating: false,
        translateLanguageCode: "en",
        reactionsSourceFlow: "businessPageReviewSection",
        minConfidenceLevel: "HIGH_CONFIDENCE",
        highlightType: "",
        highlightIdentifier: "",
        isHighlighting: false,
      },
      extensions: {
        operationType: "query",
        documentId: "ef51f33d1b0eccc958dddbf6cde15739c48b34637a00ebe316441031d4bf7681",
      },
    },
  ]);
  const headers = {
    accept: "*/*",
    "accept-language": "en-US,en;q=0.9",
    "cache-control": "no-cache",
    "content-type": "application/json",
    origin: "https://www.yelp.com",
    referer: url,
    "x-apollo-operation-name": "GetBusinessReviewFeed",
  };
  return await postJson("https://www.yelp.com/gql/batch", payload, headers);
}

export async function scrapeReviews(url, maxReviews = null) {
  const businessHtml = await fetchRenderedHtml(url, 'meta[name="yelp-biz-id"]');
  const businessId = parseBusinessId(businessHtml);
  if (!businessId) return [];

  const first = await requestReviewsApi(url, 1, businessId);
  const firstData = parseReviewData(first);
  const reviews = firstData.reviews;
  let total = firstData.total_reviews;
  if (maxReviews && maxReviews < total) total = maxReviews;

  for (let offset = 11; offset < total; offset += 10) {
    try {
      const resp = await requestReviewsApi(url, offset, businessId);
      reviews.push(...parseReviewData(resp).reviews);
    } catch (e) {
      // continue
    }
  }
  return reviews;
}

function makeSearchUrl(keyword, location, offset) {
  const params = new URLSearchParams({ find_desc: keyword, find_loc: location, start: String(offset) });
  return `https://www.yelp.com/search?${params.toString()}`;
}

export async function scrapeSearch(keyword, location, maxPages = null) {
  const firstHtml = await fetchRenderedHtml(
    makeSearchUrl(keyword, location, 0),
    "script[data-id='react-root-props']",
  );
  const data = parseSearch(firstHtml);
  const searchData = data.search_data;
  const totalResults = data.total_results;
  let totalPages = Math.max(1, Math.ceil(totalResults / 10));
  if (maxPages && maxPages < totalPages) totalPages = maxPages;

  for (let offset = 10; offset < totalPages * 10; offset += 10) {
    try {
      const html = await fetchRenderedHtml(
        makeSearchUrl(keyword, location, offset),
        "script[data-id='react-root-props']",
      );
      searchData.push(...parseSearch(html).search_data);
    } catch (e) {
      // continue
    }
  }
  return searchData;
}
